### ISSUE TITLE
RPC curl request format

### ISSUE THREAD

**Issue by georgearchon** [neo-project/neo-cli#313]
In using curl requests to interact with a smart contract I have deployed to testnet, I am encountering
many errors related to the format of my curl request including: `{"jsonrpc":"2.0","id":1,"error":{"code":-2147467261,"message":"Object reference not set to an instance of an object."}}`

My smart contract is written in C# and the Main function signature looks like:
`public static bool Main(string operation, params object[] args)`
where the args object is expected to have entries with types in order `string, byte[]`

The way I am forming my curl request is:
`curl -v -d '{"jsonrpc": "2.0","method": "invoke","params": ["<script hash>",[{"type":"String","value":"exampleOperation"},"params":{{"type":"String","value":"helloExample"},{"type":"ByteArray","value":"00"}}]],"id": 1}' http://127.0.0.1:20332`

I'm not sure how I should form the request to properly interact with my contract.
Can anyone with experience with this please make suggestions?
(PS keeping in on the level of http calls to rpc for the sake of a programmable interface so please don't suggest gui type option).

Thanks so much in advance!

**Comment by georgearchon** [comment]
Another error I've been getting in trying to get the request format down is:
`{"jsonrpc":"2.0","id":1,"error":{"code":-2147467262,"message":"Unable to cast object of type \u0027Neo.IO.Json.JObject\u0027 to type \u0027Neo.IO.Json.JArray\u0027."}}`
where the attempted request is:
`curl -v -d '{"jsonrpc": "2.0","method": "invoke","params": ["<scripthash>",{"type":"String","value":"someOperationName"},["example123",00]],"id": 1}' http://127.0.0.1:20332`

**Comment by georgearchon** [comment]
This request appears does not produce errors, but the return value from the contract is incorrect:
`curl -v -d '{"jsonrpc": "2.0","method": "invokefunction","params": ["<scripthash>","Main",[{"type":"String","value":"someOperationName"},{"type":"Array", "value":[{"type":"String","value":"example123"},{"type":"ByteArray","value":"00"}]}]],"id": 1}' http://127.0.0.1:20332`

The return value from the contract is supposed to be bool false but instead it is:
`{"jsonrpc":"2.0","id":1,"result":{"script":"<script>","state":"HALT, BREAK","gas_consumed":"0.059","stack":[{"type":"ByteArray","value":""}],"tx":"<tx>"}}`

Does this mean my request was not well-formed? What else could be the issue?

**Comment by vncoelho** [comment]
Hi @geogearchon, I am quite busy today but I will try to check your case.

Meanwhile, try to compile, deploy and invoke at https://neocompiler.io

**Comment by georgearchon** [comment]
@vncoelho I just looked at the tool https://neocompiler.io, and it says the 'deploy' feature to the Neo testnet is incomplete.

**Comment by vncoelho** [comment]
Use it normally, we have a shared privatenet that resets 12 hours.
The testnet can be used but our system was not designed based on security, the focus was a didactic plataform.

**Comment by georgearchon** [comment]
As stated in the first post my goal is to make http calls to the rpc for I am programming against that endpoint, so a gui/webui is not useful this point in my project.

**Comment by jsolman** [comment]
> This request does not produce errors, but the return value from the contract is incorrect:
curl -v -d '{"jsonrpc": "2.0","method": "invokefunction","params": ["<scripthash>","Main",[{"type":"String","value":"someOperationName"},{"type":"Array", "value":[{"type":"String","value":"example123"},{"type":"ByteArray","value":"00"}]}]],"id": 1}' http://127.0.0.1:20332


You are not supposed to include the types in your request. Try something like:

curl -v -d '{"jsonrpc": "2.0","method": "invokefunction","params": ["<scripthash>","Main",[“someOperationName",["example123", "00"]]}' http://127.0.0.1:20332

**Comment by jsolman** [comment]
Sorry, mistapped on mobile and accidentally closed and reopened.

**Comment by georgearchon** [comment]
Ha. Indecisive.

**Comment by jsolman** [comment]
> The way I am forming my curl request is:
curl -v -d '{"jsonrpc": "2.0","method": "invoke","params": ["<script hash>",[{"type":"String","value":"exampleOperation"},"params":{{"type":"String","value":"helloExample"},{"type":"ByteArray","value":"00"}}]],"id": 1}' http://127.0.0.1:20332

The operation shouldn’t have the type, and you have a jobject where it should have a JArray. Try formatting like this instead:
```
curl -v -d '{"jsonrpc": "2.0","method": "invoke","params": ["<script hash>","exampleOperation",[{"type":"String","value":"helloExample"},{"type":"ByteArray","value":"00"}]],"id": 1}' http://127.0.0.1:20332
```

**Comment by jsolman** [comment]
See also

https://docs.neo.org/en-us/node/cli/2.7.4/api/invoke.html

https://docs.neo.org/en-us/node/cli/2.7.4/api/invokefunction.html

**Comment by georgearchon** [comment]
Thanks for the last two tips @jsolman . From that formatting of the curl request I was able to invoke a function belonging to a 'helloworld'-ish contract I deployed to the test net. 
Now I am encountering another small problem:
I cannot retrieve a value stored at a contract. I'd like to understand what I am doing wrong.
I deployed this contract with storage to the test net with script hash 0x0f8236169fc023893ce7f6882aae9157162b1fac
I made an invoke function call of the single method of the contract which should have stored a key value pair <inputstring>:"boo" where in my invocation the input string is "hello".

```
curl -v -d '{"jsonrpc": "2.0","method": "invokefunction","params": ["0x0f8236169fc023893ce7f6882aae9157162b1fac","store",[{"type":"String","value":"hello"}]],"id": 1}' http://127.0.0.1:20332

```

The invocation returned what appears to be a good return value which is a hex that translates to "stored". 
```
{"jsonrpc":"2.0","id":1,"result":{"script":"0568656c6c6f51c10573746f726567ac1f2b165791ae2a88f6e73c8923c09f1636820f","state":"HALT, BREAK","gas_consumed":"1.075","stack":[{"type":"ByteArray","value":"73746f726564"}]}}

```
But when I call `getstorage` using the hex representation of the key, I get a null result.
```
curl -v -d '{"jsonrpc": "2.0","method": "getstorage","params": ["0x0f8236169fc023893ce7f6882aae9157162b1fac","68656c6c6f"],"id": 15}' http://127.0.0.1:20332

```
The above call appears to me to be consistent with what is outlined in the docs
https://docs.neo.org/en-us/node/cli/latest-version/api/getstorage.html

I'd like to get past these syntax mishaps so I can get into the meat of my project. All help is appreciated (but I'm keeping this low-level, please no gui suggestions).
For reference, here is the C# contract which compiled and deployed properly using dotnet and neo-compiler as well as https://neocompiler.io
```
using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;

namespace Neo.SmartContract
{
  public class Testy : Framework.SmartContract
  {
    public static object Main(string operation, params object[] args)
    {
      switch(operation) 
      {
        // args = [(string)word]
        case "store":
          return Store((string)args[0]);
        default:
          return false;
      }
    }

    private static string Store(string word)
    {
      string suprise = "boo";
      Storage.Put(Storage.CurrentContext, word, suprise);
      string returnVal = "stored";
      return returnVal;
    }
  }
}

```

**Comment by georgearchon** [comment]
.. Also compiling, deploying, and invoking the above C# contract yields null results when performing a storage query on the script hash with the invoked key all within https://neocompiler.io.
@jsolman , @vncoelho 
I'm very stuck on this. I really hope to get deployment and interaction with this contract figured out so I can move on to the meat of my project.
Any help is appreciated. Thanks in advance.

**Comment by jsolman** [comment]
We can look into your issue that uses a string key, but can you try using a byte array key to see if it works then?

**Comment by jsolman** [comment]
Oh, actually you can’t invoke function, that is only for read only operations, since it doesn’t send any transaction to the chain.

**Comment by jsolman** [comment]
Notice from the docs:

https://docs.neo.org/en-us/node/cli/2.7.4/api/invokefunction.html

“NOTE
This method is to test your VM script as if they were ran on the blockchain at that point in time. This RPC call does not affect the blockchain in any way.”

**Comment by jsolman** [comment]
You need to create a signed invocation transaction that calls your store operation and use `sendrawtransaction` to send it via rpc.

**Comment by jsolman** [comment]
FYI: You will need to use a library such as neon-js to generate the invocation transaction and sign it with a private key. You can’t easily do it with just a command line and curl. `neo-cli` has basic support now for creating an invocation transaction and invoking your smart contract. Currently it only supports passing string parameters, but more support for other parameter types will be added soon.

**Comment by georgearchon** [comment]
Thanks @jsolman for that info. I tried earlier to change the string key param to a byte param and I encountered the same difficulty both with the cli and the http://neocompiler.io tool. 
I will look into neon-js to generate the invocation tx and sign with a private key.
The component of my project that is interacting with the Neo-Smart contract is a server written in C/C++ (for the sake of performance). I will read through the source of neon-js to make sense of the invocation/signing steps to evaluate whether the routine (C/C++ server) > (neon-js) > (Neo SC)
could be simplied to (C/C++ server) > (Neo SC) by writing the invocation/signing steps in C/C++ and absorbing it into my server code.

Thanks for looking into this!
